### PR TITLE
fix(opencode): handle Windows project basenames

### LINKED ISSUE #652: fix(opencode): compaction recovery emits absolute Windows path as project
### 📋 Pre-flight Checks

- [x] I have searched [existing issues](https://github.com/Gentleman-Programming/engram/issues) and this is not a duplicate
- [x] I understand this issue needs `status:approved` before a PR can be opened

### 📝 Bug Description

The OpenCode plugin's project-name fallback is not Windows-safe. When OpenCode starts from a non-Git Windows directory such as `C:\Users\Blackie`, `extractProjectName()` falls back to `directory.split("/").pop()`. Because the directory contains backslashes, the function returns the entire absolute path instead of the basename.

The compaction hook then injects this invalid instruction into the compacted summary:

```text
FIRST ACTION REQUIRED: Call mem_session_summary with the content of this compacted summary. Use project: 'C:\Users\Blackie'.
```

Following that instruction causes `mem_session_summary` to reject the project because Engram requires a project name, not a filesystem path.

The same slash-only expression is also used for `oldProject`:

```ts
const oldProject = ctx.directory.split("/").pop() ?? "unknown"
```

This report is related to #583 and #594, but it is not the same defect: those issues cover session-aware or explicit project routing. This report covers the OpenCode adapter emitting an absolute Windows path as the project value before the MCP write is attempted.

### 🔄 Steps to Reproduce

1. On Windows, start OpenCode from a non-Git directory such as `C:\Users\Blackie`.
2. Use the Engram OpenCode plugin and trigger session compaction.
3. Inspect the generated `FIRST ACTION REQUIRED` instruction at the top of the compacted summary.
4. Observe that it says `Use project: 'C:\Users\Blackie'` rather than a project name.
5. Call `mem_session_summary` with the generated `project` value.
6. Observe the `invalid_project` error.

### ✅ Expected Behavior

The OpenCode adapter should derive a valid project name using a cross-platform path operation. At minimum, the final fallback for `C:\Users\Blackie` should be `Blackie`, never the absolute path.

The generated compaction recovery instruction must contain a value accepted by `mem_session_summary`.

### ❌ Actual Behavior

The adapter emits the full Windows path and Engram rejects it:

```json
{
  "error_code": "invalid_project",
  "hint": "Use a non-empty project name, not a path.",
  "message": "Project \"C:\\\\Users\\\\Blackie\" is invalid: project must be a name, not a path"
}
```

Relevant source on current `main`:

```ts
function extractProjectName(directory: string): string {
  // ... git remote and git-root attempts ...
  return directory.split("/").pop() ?? "unknown"
}

export const Engram: Plugin = async (ctx) => {
  const oldProject = ctx.directory.split("/").pop() ?? "unknown"
  const project = extractProjectName(ctx.directory)
  // ...
}
```

### Operating System

Windows

### Engram Version

1.20.0

### Agent / Client

OpenCode

### 📋 Relevant Logs

```shell
engram_mem_session_summary [project=C:\Users\Blackie, content=...]

{"available_projects":null,"error_code":"invalid_project","hint":"Use a non-empty project name, not a path.","message":"Project \"C:\\\\Users\\\\Blackie\" is invalid: project must be a name, not a path"}
```

### 💡 Additional Context

- `plugin/opencode/engram.ts` and its generated copy at `internal/setup/plugins/opencode/engram.ts` currently contain the slash-only fallback.
- A cross-platform basename implementation should be covered by a Windows-path regression test.
- The first local recovery attempt also supplied an arbitrary unregistered `session_id` and correctly failed with `unknown_session`; that separate caller mistake is intentionally excluded from this report. The reproducible adapter bug is the invalid absolute project path.

### PR DESCRIPTION
**Pull request by dnlrsls**
## Linked Issue

Closes #652

## PR Type

- [x] `type:bug` - Bug fix
- [ ] `type:feature` - New feature
- [ ] `type:docs` - Documentation only
- [ ] `type:refactor` - Code refactoring (no behavior change)
- [ ] `type:chore` - Maintenance, dependencies, tooling
- [ ] `type:breaking-change` - Breaking change

## Summary

- Derive OpenCode project basenames from both Windows and POSIX separators.
- Apply the same basename logic to Git-root fallback, cwd fallback, and project migration.
- Add runtime regressions for Windows compaction, migration, and preserved POSIX behavior.

## Changes

| File | Change |
|------|--------|
| `plugin/opencode/engram.ts` | Add separator-agnostic project basename handling. |
| `plugin/opencode/engram.test.ts` | Add Windows and POSIX runtime regression coverage. |
| `internal/setup/plugins/opencode/engram.ts` | Synchronize the embedded OpenCode plugin copy. |

## Test Plan

- [ ] Unit tests pass locally: `go test ./...`
- [x] E2E tests pass locally: `go test -tags e2e ./internal/server/...`
- [x] Manually tested the affected functionality

Additional verification:

- `node --test plugin/opencode/engram.test.ts` (3 passed)
- `node --check plugin/opencode/engram.ts`
- `node --check internal/setup/plugins/opencode/engram.ts`
- `go test ./internal/setup -run OpenCode -count=1`
- Canonical and embedded plugin copies have identical SHA-256 hashes.

`go test ./...` was attempted on Windows. It still hits unrelated existing Windows-specific failures and a `cmd/engram` timeout; the affected OpenCode-focused tests pass, and Linux CI remains authoritative for the full suite.

## Automated Checks

| Check | What it verifies | Status |
|-------|-----------------|--------|
| **Check Issue Reference** | PR body contains `Closes #N` / `Fixes #N` / `Resolves #N` | Pending |
| **Check Issue Has status:approved** | Linked issue has `status:approved` label | Pending |
| **Check PR Has type:* Label** | PR has exactly one `type:*` label | Pending |
| **Unit Tests** | `go test ./...` passes | Pending |
| **E2E Tests** | `go test -tags e2e ./internal/server/...` passes | Pending |

## Contributor Checklist

- [x] I linked an approved issue above (`Closes #N`)
- [x] I added exactly one **type:** label to this PR
- [ ] I ran unit tests locally: `go test ./...`
- [x] I ran e2e tests locally: `go test -tags e2e ./internal/server/...`
- [x] Docs updated (not required; no user-facing workflow changed)
- [x] Commits follow conventional commits format
- [x] No `Co-Authored-By` trailers in commits

## Notes for Reviewers

The bounded reliability review approved this candidate with two non-blocking warnings: trailing directory separators are not normalized, and the TypeScript regression file is not yet registered in repository CI. Both are intentionally left as follow-up scope rather than changing the frozen candidate after approval.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved project name detection for Windows-style paths by correctly extracting the last non-empty directory segment across both `/` and `\`.
  * Made migration context derivation consistent during compaction recovery, ensuring accurate `old_project`/`project` values and correct `/projects/migrate` request behavior (including skipping migration when unchanged).
* **Tests**
  * Added coverage for `experimental.session.compacting`, validating request payloads and triggering conditions across Windows/POSIX paths and edge cases (e.g., trailing separators, nested directories, and Git-root-derived basenames).

### CODE REVIEW

File: plugin/opencode/engram.test.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  const output = { context: [] as string[] }
+
+  await plugin["experimental.session.compacting"]?.(
+    { sessionID: "session-652" } as never,
+    output,
+  )
+
+  assert.match(output.context.at(-1) ?? "", /Use project: 'Blackie'/)
+  assert.doesNotMatch(output.context.at(-1) ?? "", /C:\\Users\\Blackie/)
+})
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply plugin/opencode/engram.test.ts:52]: Added. The test now makes remote lookup fail, returns a Windows path from rev-parse --show-toplevel, and verifies the compaction project uses the Git-root basename.

File: internal/setup/plugins/opencode/engram.ts [hunk 2]
```diff
@@ -153,6 +153,10 @@ async function isEngramRunning(): Promise<boolean> {
 
 // ─── Helpers ─────────────────────────────────────────────────────────────────
 
+function basename(path: string): string {
+  return path.split(/[\\/]/).filter(Boolean).pop() ?? "unknown"
+}
+
```
[the remark this answers is not in the document]
Comment (dnlrsls) [reply internal/setup/plugins/opencode/engram.ts:160]: Fixed in d0ee245. Both plugin copies reject bare drive labels and return unknown for C:\\ and C:/; the regression tests also verify the exact legacy migration keys for both forms.

**Comment by dnlrsls** [comment]
Closing as superseded. Current main resolves the OpenCode project identity through /project/current, backed by project.DetectProjectFull and platform-aware filepath.Base normalization; coverage is in plugin/opencode/engram.test.mjs. This PR's local basename extraction would replace that canonical resolver, and its legacy {old_project,new_project} migration payload is rejected by the current server contract. No tests were run for this closure.
